[Zeke]Step2 - 음료수 클래스 계층과 자판기 객체의 행위, XCTest

zekexros opened the pull request:
- 작업 목록
  - 음료수
    - [x] 음료수 클래스 계층을 최상위 객체부터 3단계 이상 구분
    - [x] 상속받은 하위 클래스에도 상위에 없는 속성 한가지 이상 추가
    - [ ] 추가한 속성에 대한 인터페이스 작성
  - 자판기
    - [x] 원하는만큼 금액 추가하는 메소드
    - [x] 재고 추가 메소드
    - [x] 현재금액으로 구매가능 음료수 리스트 반환 메소드
    - [x] 음료수 구매 메소드
    - [x] 잔액 확인 메소드
    - [x] 종류별로 재고 반환하는 메소드
    - [x] 유통기한 지난 재고 반환하는 메소드
    - [x] 재고중에 따뜻한 음료수 반환하는 메소드
    - [x] 구매상품 이력 반환하는 메소드

 


- 학습 키워드

  - designated init
  - convenience init
  - XCTest
  - 스위프트의 프로토콜과 자바의 인터페이스 차이
  - Hashable 프로토콜
  - Equatable 프로토콜
  
  


- 고민과 해결

  - 전체 상품 재고를 종류별로 반환하기 위해 `Dict`타입을 이용해야 한다. `Dict`타입의 key타입에 Beverage를 대입한 순간 컴파일 에러가 나왔다. 이유인 즉, `Dict`타입의 key타입은 `Hashable`을 채택한 타입만 들어갈 수 있었다. 따라서 Beverage타입에 `Hashable`프로토콜을 채택하였지만 다시 컴파일 에러가 등장하였다. stackoverflow의 한 질의응답을 통해 hash()메소드를 통해 hash값을 제공해야한다는것을 알고 객체의 속성들을 제공하여 문제를 해결하였다.
  - Beverage는 `Hashable`을 사용하기 위해선 `Equatable`프로토콜 또한 채택해야한다. 이유는 `Hashable` 프로토콜은 `Equatable` 프로토콜에서 상속되기 때문에 요구 사항이 충족되어야 한다. 따라서 프로토콜을 따라 타입메소드 '=='를 description 비교를 반환하여 구현하였다.
  



- 부연 설명

  - Drinks클래스 안에 Beverage타입을 배열로 갖는 purchasedDrinks라는 프로퍼티를 생성하여 구매했던 음료수들의 이력을 반환하는 메소드를 구현하였다.
- 질문거리
  - '추가한 속성에 대한 인터페이스 작성'이라는 것은 음료수 및 하위타입들에 대해 개별적으로 프로토콜을 하나씩 생성하는건가요??

zekexros committed:
date속성의 이름 변경

date -> manufactureOfDate로 변경

zekexros committed:
Date타입의 anotherDescription 속성 추가

기존 Date타입의 description의 기능을 보존하기위해 anotherDescription이름의 새로운 속성 생성

zekexros committed:
printDrinks() 메소드 이름 변경

상위타입에서 하위타입의 속성들을 클로저를 이용하여 사용하는데 변수명에 print가 포함되어있는것은 적절치 않아 변경

zekexros committed:
Coffee, TOP객체의 속성과 인터페이스 추가

Coffee에서 카페인함유량 속성, 고카페인여부 인터페이스 추가하고 TOP에서 맛속성, 에스프레소맛여부 인터페이스 추가

zekexros committed:
CashManagementSystem 생성

CashManagementSystem 생성 및 메소드 생성

zekexros committed:
2,3단계 클래스에 속성과 인터페이스 추가

우유, 탄산, 에너지드링크, 커피 클래스와 그 하위타입들에 속성과 인터페이스 추가

zekexros committed:
구매가능목록 반환 메소드 추가

현재 금액으로 구매가능한 음료수 목록을 리턴하는 메소드 추가

zekexros committed:
자판기 금액 조절 메소드 추가

입력값을 Int타입으로 받아 금액을 조절하는 메소드 추가

zekexros committed:
음료수 구매 메소드 추가

음료수를 구매하게되면 재고에서 음료수를 제거하고 금액도 수정하는 메소드를 구현

zekexros committed:
잔액을 확인하는 메소드 추가

CashManagementSystem 객체의 cash속성을 가져오는 메소드 구현

zekexros committed:
재고를 반환하는 메소드 추가

Beverage에 hashable을 적용하여 Dict타입의 keytype에 사용할수있게되어 재고 종류와 개수를 Dict타입으로 반환

zekexros committed:
유통기한 지난 재고 반환하는 메소드 추가

Beverage에 유통기한 지난 여부를 반환하는 메소드를 구현하고 상위타입에서 이용하여 구현

zekexros committed:
Beverage의 속성 추가

Beverage에 유통기한과 온도 속성을 새로 추가

zekexros committed:
따뜻한음료 반환하는 메소드 추가

음료 객체의 온도속성을 이용하여 따뜻한음료를 반환하는 메소드 구현

zekexros committed:
구매 상품 이력 반환하는 메소드 추가

Drinks클래스에 purchasedDrinks 배열형태의 프로퍼티를 추가하여 구매 상품 이력을 반환하는 메소드 구현

zekexros committed:
XCTest 추가

XCTest를 생성하여 타겟 지정

zekexros committed:
테스트 시나리오 추가

XCTest에 테스트 시나리오를 따라 구현하고 동작을 확인

zekexros committed:
README.md 수정

step2 반영

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -9,20 +9,53 @@ import Foundation
     private var name: String
-    private var date: Date
-    init(brand: String, capacity: Int, price: Int, name: String, date: Date) {
+    private var dateOfManufacture: Date
+    private var expirationDate: Date
```
속성 이름에 타입 이름을 그대로 붙이지 않아도 충분히 유추할 수 있습니다. 
Date를 과감하게 빼보세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -9,20 +9,53 @@ import Foundation
+        self.expirationDate = expirationDate
+        self.temperature = temperature
+    }
+    
+    func isAvailablePurchase(with cash: Int) -> Bool{
+        return self.price <= cash
+    }
+    func calculateChange(with cash: Int) -> Int {
+        return cash - self.price
+    }
+    func isExpired(on standard: Date) -> Bool {
+        return self.expirationDate < standard
+    }
+    func isHot(at standard: Double) -> Bool {
+        return self.temperature >= standard
     }
```
여기 메소드들 사이는 줄 간격이 떨어지지 않았네요. 의도한 것일까요?
빈공백이나 줄도 일관성 있게 작성해주는 게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -9,20 +9,53 @@ import Foundation
     }
+}
 
+extension Beverage: Equatable {
+    static func == (lhs: Beverage, rhs: Beverage) -> Bool {
+        return lhs.description == rhs.description
+    }
+}
+
+extension Beverage: Hashable {
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(brand)
+        hasher.combine(capacity)
+        hasher.combine(price)
+        hasher.combine(name)
+        hasher.combine(dateOfManufacture)
+    }
 }
```
- Equatable과 Hashable 은 어떤 차이가 있는건가요? 
- 특히 description을 로직 구현하는 데 사용하지마세요. 나쁜 습관입니다. 속성값과 비교하는 게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/CashManagementSystem.swift:
```diff
@@ -0,0 +1,34 @@
+    
+    func recharge(with cash: Int) {
+        self.cash = cash
+    }
+    func showListForPurchase(with drinks: Drinks) ->[Beverage] {
+        return drinks.showListForPurchase(with: cash)
+    }
+    
+    func checkBalance() -> Int{
+        return self.cash
+    }
+    
+    func isAvailableForPurchase(with beverage: Beverage) -> Bool {
+        beverage.isAvailablePurchase(with: self.cash)
+    }
+    
+    func increaseCash(with amount: Int) {
+        self.cash += amount
+    }
+    func decreaseCash(with beverage: Beverage) {
+        self.cash = beverage.calculateChange(with: cash)
+    }
```
여기 선언한 함수들 시그니처에 줄간격이 들쭉날쭉 합니다. 사소한 거지만 가독성을 떨어뜨립니다. 일관성 있게 작성하세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/CashManagementSystem.swift:
```diff
@@ -0,0 +1,34 @@
+
+import Foundation
+
+class CashManagementSystem {
+    private var cash: Int = 0
+    
+    func recharge(with cash: Int) {
+        self.cash = cash
+    }
+    func showListForPurchase(with drinks: Drinks) ->[Beverage] {
+        return drinks.showListForPurchase(with: cash)
+    }
+    
+    func checkBalance() -> Int{
+        return self.cash
+    }
```
이정도 타입이면 내부 cash를 return 하지 않고 그대로 (이름은 좀 거창하지만) CashManagementSystem을 타입으로 사용해도 좋을 것 같습니다. 세부적인 타입을 만들면 그대로 쓰려고 노력해보세요.

zekexros replied:
내부의 cash를 return하지 않고 그대로 타입을 사용한다. 문장을 계속 읽고있지만 어떻게해야할지 감이 잘 안옵니다🥲 checkBalance()의 반환타입을 CashManagementSystem으로 하고 맞춰서 구현하면될까요?

godrm replied:
네 말하자면 그런거죠. 그러면 그 타입을 그대로 다른 곳에서도 사용하고, 메소드로 접근하게 될테니까요.
카드 게임에서 Rank나 Suit를 빼서 넘겨주는 게 아니라 Card 를 넘겨주는 것처럼요

godrm commented on VendingMachineApp/VendingMachineApp/Model/CashManagementSystem.swift:
```diff
@@ -0,0 +1,34 @@
+        self.cash = cash
+    }
+    func showListForPurchase(with drinks: Drinks) ->[Beverage] {
+        return drinks.showListForPurchase(with: cash)
```
이 메소드는 좀 오버같습니다. Drinks에 cash를 넘기는 게 어떨까요? 
잔돈 시스템에 음료수 목록을 전달하는 게 어색합니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Coffee/Coffee.swift:
```diff
@@ -8,7 +8,12 @@
+        super.init(brand: brand, capacity: capacity, price: price, name: name, dateOfManufacture: dateOfManufacture, expirationDate: expirationDate, temperature: temperature)
+    }
+    func isHighCaffeine() -> Bool {
+        return self.caffeineContent >= 0.15
```
이런 비교식은 기준값을 상수로 선언하고 매개변수로 전달하는 게 좋습니다.
특히 0.15 같은 Double 실수값을 비교할 때는 오차를 같이 고려하셔야 합니다. 
특정한 값에 따라서 같을수도 다를 수도 있습니다. 오차 범위에 대해 학습해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Drinks.swift:
```diff
@@ -9,18 +9,45 @@ import Foundation
 
 class Drinks {
     private var drinks: [Beverage]
+    private var purchasedDrinks: [Beverage]
```
얼핏 보면 이 두 가지를 모두 쓰는 메소드보다 하나씩 쓰는 메소드가 더 많은 거 아닐까요?
두 속성이 얼마나 관련이 있을까요?

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -9,16 +9,44 @@ import Foundation
+        return cashManagementSystem.checkBalance()
+    }
+    
+    func showStock() -> [Beverage:Int] {
+        drinks.giveStockList()
+    }
+    func showExpiredStock() -> [Beverage] {
+        drinks.giveExpiredList()
+    }
+    func showHotBeverageList() -> [Beverage] {
+        drinks.giveHotBeverageList()
+    }
+    func showPurchasedList() -> [Beverage] {
+        drinks.givePurchasedList()
     }
```
전체적으로 자판기 코드는 매우 깔끔하네요.
여기서 리턴하는 [Beverage] 대신에 Drinks를 사용할 수는 없는건가요? 어떤 차이가 있을까요?

godrm reviewed (changes requested):
수고하셨습니다. 다음 피드백을 확인해보세요.

zekexros committed:
dateOfManufacture 이름 변경

기존 dateOfManufacture에서 manufacture로 변경

zekexros committed:
줄바꿈, 줄 간격 수정

일관성있게 규칙적으로 메소드간 줄바꿈, 함수들의 시그니처에 줄간격 수정

zekexros committed:
hashable 프로토콜 수정

equatable프로토콜을 상속받는 hashable에 '=='타입메소드를 구현하여 수정

zekexros committed:
CashManagementSystem객체의 행위 이전

구매가능목록을 반환하는 행위를 CashManagementSystem객체에서 Drinks객체로 이전

zekexros committed:
showListForPurchase() 수정

메소드 안에 return 추가

zekexros committed:
isHighCaffeine() 수정

기준값을 매개변수로 전달하는것, 오차범위를 소수점 2자리까지 고려하도록 수정

zekexros committed:
PurchasedList 분리 및 클래스 생성

기존 Drinks객체로부터 분리하여 새로 PurhcasedList클래스 생성

zekexros committed:
자판기의 메소드 리턴 타입 변경

리턴 타입을 기존 [Beverage] 에서 Drinks로 변경

godrm commented:
@zeke-iOS 수정이 끝난건가요? 확인해서 머지하면 되나요?

zekexros committed:
extension으로 equatable 채택

XCTest에서 비교값으로 사용하기 위해 Drinks, CashManagementSystem, PurchasedList에서 Equatable 채택

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -9,53 +9,53 @@ import Foundation
     private var name: String
-    private var dateOfManufacture: Date
+    private var manufacture: Date
     private var expirationDate: Date
```
`manufacture`는 날짜 느낌이 부족하니까 manufacturedAt 처럼 표현해도 좋을 것 같습니다. 
expirationDate도 개선해보세요.

zekexros committed:
checkBalance() 수정

매개변수에 대입된 0을 cash속성으로 변경

zekexros commented:
수정하였습니다😀

## 바뀐것

- dateOfManufacture의 이름 변경 -> 다시 변경하겠습니다!!

- 줄바꿈, 줄 간격 수정 -> XCode를 화면의 반만 사용하다보니 띄어쓰기가 눈에 잘 안띄었네요 유의하겠습니다.

- hashable 프로토콜 수정 -> Equatable을 상속받으므로 hashable안에 '=='메소드 구현

- CashManagementSystem 객체에 있던 showListForPurchase()는 어색하다는것을 확인하였습니다. 따라서 Drinks로 옮겼습니다.

- isHighCaffeine() 수정 -> 소수점 2자리로 제한하는것이 오차 범위를 줄이는것이라고 생각하였습니다. 이러한 방향이 맞는지 사실 잘 모르겠습니다.

- PurchasedList Class생성 -> 기존에 Drinks에 들어있었던 속성이지만 메소드에서 같이 사용되지 않아 연관성이 없으니 따로 class로 분리하였습니다.

- 자판기의 메소드들은 [Beverage]타입으로 반환되었는데, 이는 활용성이 너무나 떨어진다. 생성한 세부타입을 적극적으로 이용하지 못한다. 따라서 기존 생성한 Drinks, PurchasedList, CashManagementSystem 타입으로 반환하게 하였습니다.

- 반환타입이 바뀜에 따라 XCTest도 수정해야한다. 새로 생성한 타입들에 Equatable을 채택하여 XCTAssertEqual을 이용할 수 있게 되었습니다.

godrm commented:
네. 다음 단계를 진행해봅시다.
